### PR TITLE
Remove task templates if required features are removed during `Dataset.map`

### PR DESCRIPTION
**Pull request by lewtun**
This PR fixes a bug reported by @craffel where removing a dataset's columns during `Dataset.map` triggered a `KeyError` because the `TextClassification` template tried to access the removed columns during `DatasetInfo.__post_init__`:

```python
from datasets import load_dataset

# `yelp_polarity` comes with a `TextClassification` template
ds = load_dataset("yelp_polarity", split="test")
ds
# Dataset({
#     features: ['text', 'label'],
#     num_rows: 38000
# })

# Triggers KeyError: 'label' - oh noes!
ds.map(lambda x: {"inputs": 0}, remove_columns=ds.column_names)
```

I wrote a unit test to make sure I could reproduce the error and then patched a fix.

### CODE REVIEW

File: src/datasets/arrow_dataset.py [hunk 1]
```diff
@@ -1997,6 +1997,11 @@ def init_buffer_and_writer():
         if update_data:
             # Create new Dataset from buffer or file
             info = self.info.copy()
+            # Remove task templates if the required features have been removed
+            if info.task_templates:
+                for idx, task in enumerate(info.task_templates):
+                    if not all(k in writer._features.keys() for k in task.features.keys()):
+                        _ = info.task_templates.pop(idx)
```
Comment (SBrandeis) [inline src/datasets/arrow_dataset.py:2004]: Using  `pop` here seems dangerous - what if we remove two templates? The index of the second one is not `idx`

Comment (SBrandeis) [reply src/datasets/arrow_dataset.py:2004]: Personnally, I would have used a list comprehension here:

```python
info.task_templates = [template if all(k in writer._features.keys() for k in template.features.keys()) for template in info.task_templates]
```

**Review by SBrandeis — CHANGES_REQUESTED** [review]

File: src/datasets/arrow_dataset.py (hunk 1 above)
Comment (lewtun) [reply src/datasets/arrow_dataset.py]: oh yes you are indeed correct - well spotted! i've now switch to using a list comprehension as you suggested

**Review by SBrandeis — APPROVED** [review]
LGTM :)

**Review by lhoestq — APPROVED** [review]
Thanks for reporting @craffel and for fixing @lewtun :)

**Pull request merged**